pkg/compiler/analysis.go

```diff
@@ -348,9 +413,147 @@ func (c *codegen) analyzeFuncUsage() funcUsage {
 		return nil
 	}
 
+	// pickGlobalVarsFromNodes searches for global variables used in the given set
+	// of nodes calling markAsUsed for each global variable. Be careful while
+	// using codegen after pickGlobalVarsFromNodes, it changes import map, currPkg
+	// and typeInfo.
+	pickGlobalVarsFromNodes := func(nodes []nodeContext, markAsUsed func(name string)) {
+		for len(nodes) != 0 {
+			var nextExprToCheck []nodeContext
+			for _, val := range nodes {
+				// Set variable context for proper name extraction.
+				c.importMap = val.importMap
+				c.currPkg = val.currPkg
+				c.typeInfo = val.typeInfo
+				ast.Inspect(val.node, func(node ast.Node) bool {
+					switch n := node.(type) {
+					case *ast.KeyValueExpr: // var _ = f() + CustomInt{Int: Unused}.Int + 3 => mark Unused as "used".
+						nextExprToCheck = append(nextExprToCheck, nodeContext{
+							node:      n.Value,
+							path:      val.path,
+							importMap: c.importMap,
+							typeInfo:  c.typeInfo,
+							currPkg:   c.currPkg,
+						})
```

Comment:
If I get it right, we append `val` with a changed `node` field, can we move creation of such node to a separate function to make the code more vertically-compact?

Reply:
Also, did you consider using DFS (via recursive invocations) here?

Reply:
Not sure I understand what do you mean, could you, please, clarify? I don't change the `node` field, I just add it to the list of expressions we need to check next (against call containing).

> Also, did you consider using DFS (via recursive invocations) here?
Could you give an example?

Reply:
> could you, please, clarify?

I mean the appended item is constructed similarly: it just copies 4 fields and sets another one. It could be abstracted as a `append(nextExprToCheck, val.derive(n.Value)` naming can be better.

>Could you give an example?

Instead of `append(nextExprToCheck, nodeContext{node: n.Value, ...}` you can call some function recursively like `Visit(n.Value)`.

Reply:
> appended item is constructed similarly

You're right, fixed. Also, I've refactored it a bit and moved `pickGlobalVarsFromNodes` to a separate codegen-binded method.

> you can call some function recursively like `Visit(n.Value)`

It was like this before, but IMO, the code looks cleaner with `Inspect` and `nextExprToCheck` list when traversal occurs stage-by-stage. Could you review the updated code one more time and tell whether we still need to use `Visit`?

Reply:
It is not that we _need_ it, but that at the first glance it could be simpler. This is subjective, of course, I have no strong preference here.